torch/csrc/autograd/engine.cpp

```diff
@@ -112,134 +122,132 @@ auto Engine::thread_on_exception(FunctionTask& task, std::exception& e) -> void
   }
 }
 
-static variable_list call_pre_hooks(Function& fn, variable_list grad_output) {
+static variable_list call_pre_hooks(Function& fn, variable_list inputs) {
   for (auto& hook : fn.pre_hooks) {
-    grad_output = (*hook)(grad_output);
+    inputs = (*hook)(inputs);
   }
-  return grad_output;
+  return inputs;
 }
 
-static variable_list call_post_hooks(Function& fn, variable_list grad_input, variable_list grad_output) {
+static variable_list call_post_hooks(Function& fn, variable_list outputs, variable_list inputs) {
   for (auto& hook : fn.post_hooks) {
-    grad_input = (*hook)(grad_input, grad_output);
+    outputs = (*hook)(outputs, inputs);
   }
-  return grad_input;
+  return outputs;
 }
 
-static variable_list call_function(FunctionTask& task) {
-  auto grad_output = call_pre_hooks(*task.fn, GradBuffer::variables(std::move(task.grad)));
-  auto grad_input = task.fn->apply(grad_output);
-  return call_post_hooks(*task.fn, std::move(grad_input), std::move(grad_output));
+static std::pair<bool, variable_list> call_function(FunctionTask& task) {
+  auto& fn = *task.fn;
+  auto inputs = call_pre_hooks(fn, InputBuffer::variables(std::move(task.inputs)));
+
+  auto& function_callbacks = task.base->function_callbacks;
```

Comment:
Yeah it might seem weird but that's how hooks work:
* You can think of pre-hooks as hooks registered on incoming edges, but since we have no "edge objects", we need to register them in the next function. But they are `pre_hooks` because they always execute before the inputs even arrive at the function. Also, this is needed because we want to execute the Variable hooks even when we'll be capturing the grad using a callback.
* Post hooks are for reading/overriding the output of the function, so it doesn't make sense to call them if a callback aborts the operation.